Add cdi settings doc (#841)
harvester/docs: docs/advanced/storageclass.md

For more information, see [Data Locality](https://longhorn.io/docs/1.6.0/high-availability/data-locality/) in the Longhorn documentation.

> ## CDI Settings
> 
> Harvester integrates with Containerized Data Importer ([CDI](https://kubevirt.io/user-guide/storage/containerized_data_importer/)) to handle VM image management for supported storage classes:

Reviewer: ```suggestion
## Containerized Data Importer (CDI) Settings

Harvester integrates with the [Containerized Data Importer](https://kubevirt.io/user-guide/storage/containerized_data_importer) (CDI) to handle VM image management for the following StorageClasses:
```

> - **Longhorn v2** (Longhorn v1 is not supported)
> - **LVM storage**
> - **Other third-party storage** (UI support not yet available)

Reviewer: ```suggestion
- Longhorn V2 Data Engine
- LVM
- Third-party storage

:::note

The Harvester UI currently does not support the use of CDI with third-party storage.

:::
```

> To enable editing of CDI settings for day-2 operations, Harvester introduces 4 attributes to the StorageClass that automatically update the underlying CDI settings.
> 
> ![](/img/v1.6/storageclass/cdi-settings.png)
> 
> Each field in the above image corresponds to an annotation in the StorageClass:

Reviewer: ```suggestion
To enable editing of CDI settings for day-2 operations, Harvester provides StorageClass attributes that automatically update the underlying CDI settings.

![](/img/v1.6/storageclass/cdi-settings.png)

Each field on the **CDI Settings** screen corresponds to an annotation in the StorageClass.
```

> | Configuration | Annotation | Description | Supported Values | Example |
> |---------------|------------|-------------|------------------|---------|
> | **Volume Mode / Access Modes** | `cdi.harvesterhci.io/storageProfileVolumeModeAccessModes` | Specifies the default PVC volume mode and access modes. | JSON object with volume modes and access modes | `'{"Block":["ReadWriteOnce"]}'` |
> | **Volume Snapshot Class** | `cdi.harvesterhci.io/storageProfileVolumeSnapshotClass` | Sets the Volume Snapshot Class name to be used when taking snapshots of virtual machine images under this StorageClass. This setting only applies when using the `snapshot` clone strategy. If the user has already set `volumeSnapshotClassName` in the [`csi-driver-config`](./settings.md#csi-driver-config) setting for the corresponding provisioner, that value will be used as the default. | Valid VolumeSnapshotClass name | `lvm-snapshot`, `longhorn-snapshot` |
> | **Clone Strategy** | `cdi.harvesterhci.io/storageProfileCloneStrategy` | Defines the clone strategy to use for volumes created with VM images using this StorageClass. See the table below for available options. | `copy`: Copies blocks of data over the network<br/>`snapshot`: Clones the volume by creating a temporary VolumeSnapshot and restoring it to a new PVC<br/>`csi-clone`: Clones the volume using a CSI clone operation | `snapshot` |
> | **File System Overhead** | `cdi.harvesterhci.io/filesystemOverhead` | Specifies the percentage of filesystem overhead to consider when calculating PVC size. | Decimal value between 0 and 1, up to 3 digits | `0.05` |

Reviewer: ```suggestion
| UI Field | Annotation | Description | Supported Values | Example |
| --- | --- | --- | --- | --- |
| **Volume Mode / Access Modes** | `cdi.harvesterhci.io/storageProfileVolumeModeAccessModes` | Default PVC volume mode and access modes | JSON object with volume modes and access modes | `'{"Block":["ReadWriteOnce"]}'` |
| **Volume Snapshot Class** | `cdi.harvesterhci.io/storageProfileVolumeSnapshotClass` | VolumeSnapshotClass name to be used when taking snapshots of virtual machine images under this StorageClass. This setting applies only when you are using the `snapshot` clone strategy. If you have already configured the `volumeSnapshotClassName` in the `csi-driver-config` setting for the corresponding provisioner, that value will be used as the default. | Valid VolumeSnapshotClass name | `lvm-snapshot`, `longhorn-snapshot` |
| **Clone Strategy** | `cdi.harvesterhci.io/storageProfileCloneStrategy` | Clone strategy to be used for volumes created with VM images that use this StorageClass. | `copy`: Copies blocks of data over the network<br/>`snapshot`: Clones the volume by creating a temporary VolumeSnapshot and restoring it to a new PVC<br/>`csi-clone`: Clones the volume using a CSI clone operation | `snapshot` |
| **File System Overhead** | `cdi.harvesterhci.io/filesystemOverhead` | Percentage of filesystem overhead to be considered when calculating the PVC size. | Decimal value between 0 and 1 with a maximum of 3 digits | `0.05` |
```

Here is an example of a StorageClass YAML configuration:

```yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: lvm
  annotations:
    cdi.harvesterhci.io/storageProfileCloneStrategy: snapshot
    cdi.harvesterhci.io/storageProfileVolumeModeAccessModes: '{"Block":["ReadWriteOnce"]}'
    cdi.harvesterhci.io/storageProfileVolumeSnapshotClass: lvm-snapshot
    cdi.harvesterhci.io/filesystemOverhead: '0.05'
```

> :::note
> 
> It's recommended to let the Harvester controller synchronize and persist these storage profile configuration by utilizing the Harvester CDI annotations, instead of changing the storage profile or CDI directly.
> 
> :::

Reviewer: ```suggestion
:::caution

Avoid changing the storage profile or CDI directly. Instead, allow the Harvester controller to synchronize and persist the storage profile configuration through the use of CDI annotations.

:::
```

### Default Values

The following are the default values for Longhorn v2 and LVM storage classes:

**Longhorn V2:**
- `cdi.harvesterhci.io/storageProfileCloneStrategy`: `"copy"`
- `cdi.harvesterhci.io/storageProfileVolumeSnapshotClass`: `"longhorn-snapshot"`

**LVM:**
- `cdi.harvesterhci.io/storageProfileVolumeModeAccessModes`: `'{"Block":["ReadWriteOnce"]}'`
- `cdi.harvesterhci.io/storageProfileCloneStrategy`: `"snapshot"`
- `cdi.harvesterhci.io/storageProfileVolumeSnapshotClass`: `"lvm-snapshot"`

> For other third-party storage default values, please refer to https://github.com/kubevirt/containerized-data-importer/blob/v1.61.1/pkg/storagecapabilities/storagecapabilities.go#L35-L127

Reviewer 2: Suggest compacting the link with a meaningful keyword—for example, `CDI Storage Capability`.

> ### Default Values
> 
> The following are the default values for Longhorn v2 and LVM storage classes:
> 
> **Longhorn V2:**
> - `cdi.harvesterhci.io/storageProfileCloneStrategy`: `"copy"`
> - `cdi.harvesterhci.io/storageProfileVolumeSnapshotClass`: `"longhorn-snapshot"`
> 
> **LVM:**
> - `cdi.harvesterhci.io/storageProfileVolumeModeAccessModes`: `'{"Block":["ReadWriteOnce"]}'`
> - `cdi.harvesterhci.io/storageProfileCloneStrategy`: `"snapshot"`
> - `cdi.harvesterhci.io/storageProfileVolumeSnapshotClass`: `"lvm-snapshot"`
> 
> For other third-party storage default values, please refer to [CDI Storage Capability](https://github.com/kubevirt/containerized-data-importer/blob/v1.61.1/pkg/storagecapabilities/storagecapabilities.go#L35-L127).

Reviewer: ```suggestion
The following are the default values for the supported StorageClasses:

- Longhorn V2 Data Engine
    - `cdi.harvesterhci.io/storageProfileCloneStrategy`: `"copy"`
    - `cdi.harvesterhci.io/storageProfileVolumeSnapshotClass`: `"longhorn-snapshot"`

- LVM
    - `cdi.harvesterhci.io/storageProfileVolumeModeAccessModes`: `'{"Block":["ReadWriteOnce"]}'`
    - `cdi.harvesterhci.io/storageProfileCloneStrategy`: `"snapshot"`
    - `cdi.harvesterhci.io/storageProfileVolumeSnapshotClass`: `"lvm-snapshot"`

- Third-party storage: See [`storagecapabilities.go`](https://github.com/kubevirt/containerized-data-importer/blob/v1.61.1/pkg/storagecapabilities/storagecapabilities.go#L35-L127) in the CDI repository.
```

For more information, see [Data Locality](https://longhorn.io/docs/1.6.0/high-availability/data-locality/) in the Longhorn documentation.

## CDI Settings

Harvester integrates with CDI ([Containerized Data Importer](https://kubevirt.io/user-guide/storage/containerized_data_importer/)) to handle VM image management for supported storage classes:

- **Longhorn v2** (Longhorn v1 is not supported)
- **LVM storage**
- **Other third-party storage** (UI support not yet available)

To enable editing of CDI settings for day-2 operations, Harvester introduces 4 attributes to the StorageClass that automatically update the underlying CDI settings.

> ![](/img/v1.6/storageclass/cdi-settings.png)

Author: Good catch, many thanks!

Each field in the above image corresponds to an annotation in the StorageClass:

- **Volume Mode / Access Modes** - `cdi.harvesterhci.io/storageProfileVolumeModeAccessModes`: Specifies the default PVC volume mode and access modes.
- **Volume Snapshot Class** - `cdi.harvesterhci.io/storageProfileVolumeSnapshotClass`: Sets the Volume Snapshot Class name to be used when taking snapshots of PVCs under this StorageClass. This setting only applies when using the `snapshot` clone strategy. If the user has already set `volumeSnapshotClassName` in the `csi-driver-config` setting for the corresponding provisioner, that value will be used as the default.
- **Clone Strategy** - `cdi.harvesterhci.io/storageProfileCloneStrategy`: Defines the clone strategy to use for volumes created with VM images using this StorageClass:
  - **copy**: Copies blocks of data over the network
  - **snapshot**: Clones the volume by creating a temporary VolumeSnapshot and restoring it to a new PVC
  - **csi-clone**: Clones the volume using a CSI clone operation

> - **File System Overhead** - `cdi.harvesterhci.io/filesystemOverhead`: Specifies the percentage of filesystem overhead to consider when calculating PVC size.

Author: Thanks, table format is much better.

Here is an example of a StorageClass YAML configuration:

```yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: lvm
  annotations:
    cdi.harvesterhci.io/storageProfileCloneStrategy: snapshot
    cdi.harvesterhci.io/storageProfileVolumeModeAccessModes: '{"Block":["ReadWriteOnce"]}'
    cdi.harvesterhci.io/storageProfileVolumeSnapshotClass: lvm-snapshot
    cdi.harvesterhci.io/filesystemOverhead: '0.05'
```

:::note

> This is a one-way synchronization. If you directly update the StorageProfile or CDI settings, the changes will not be reflected in the StorageClass annotations.

Author: > In a way, I think we want to encourage users to use these annotations instead of creating and modifying the storage profiles directly. What do you think of removing https://docs.harvesterhci.io/v1.6/advanced/csidriver/#storage-profiles?

Make sens to me. c.c. @Vicente-Cheng, @WebberHuang1118, thoughts?

For more information, see [Data Locality](https://longhorn.io/docs/1.6.0/high-availability/data-locality/) in the Longhorn documentation.

## Containerized Data Importer (CDI) Settings

Harvester integrates with the [Containerized Data Importer](https://kubevirt.io/user-guide/storage/containerized_data_importer) (CDI) to handle VM image management for the following StorageClasses:

- Longhorn V2 Data Engine
- LVM
- Third-party storage

:::note

> The Harvester UI currently does not support the use of CDI with third-party storage.

Reviewer 3: ```suggestion
The Harvester UI currently does not support the use of CDI with third-party storage. Instead, apply the Harvester CDI annotations directly to the third-party storage class.
```
